src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -108,14 +108,19 @@ internal override void OnBufferAllocated(byte[] buffer)
         {
             Debug.Assert(_preallocatedOverlapped == null);
 
-            _preallocatedOverlapped = new PreAllocatedOverlapped(FileStreamValueTaskSource.s_ioCallback, this, buffer);
+            _preallocatedOverlapped = new PreAllocatedOverlapped(ValueTaskSource.s_ioCallback, this, buffer);
         }
 
-        private FileStreamValueTaskSource? CompareExchangeCurrentOverlappedOwner(FileStreamValueTaskSource? newSource, FileStreamValueTaskSource? existingSource)
+        private ValueTaskSource? CompareExchangeCurrentOverlappedOwner(ValueTaskSource? newSource, ValueTaskSource? existingSource)
             => Interlocked.CompareExchange(ref _currentOverlappedOwner, newSource, existingSource);
 
         public override int Read(byte[] buffer, int offset, int count)
-            => ReadAsyncInternal(new Memory<byte>(buffer, offset, count)).AsTask().GetAwaiter().GetResult();
+        {
+                ValueTask<int> vt = ReadAsyncInternal(new Memory<byte>(buffer, offset, count));
+                return vt.IsCompleted?
+                    vt.Result :
+                    vt.AsTask().GetAwaiter().GetResult();
```

Comment:
Is it possible to apply this same pattern in `void Write`?
https://github.com/dotnet/runtime/blob/9bebe1f4b5ab5ec63ff537140d89a166a44d1d7f/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs#L238-L239

Reply:
I tried, but a `ValueTask` does not have a `Result` property, like a `ValueTask<T>` does, because it returns void. So I'm not sure if it's possible to apply an equivalent pattern.

Reply:
```C#
public override void Write(byte[] buffer, int offset, int count)
{
    ValueTask vt = WriteAsyncInternal(new ReadOnlyMemory<byte>(buffer, offset, count), CancellationToken.None);
    if (vt.IsCompleted)
    {
        vt.GetAwaiter().GetResult();
    }
    else
    {
        vt.AsTask().GetAwaiter().GetResult();
    }
}
```
That said, it's not much of an issue for this case.  The read side is an issue because AsTask() will almost certainly need to allocate a new `Task<int>` in order to hand back the result, but for the write side, AsTask() on an already completed ValueTask will just return Task.Completed, assuming the operation completed successfully.  So the only time the existing code for the write side will allocate is when the operation was canceled or faulted, in which case we're about to throw an exception anyway, and the extra task allocation won't really matter.  Up to you whether you want to change it.